vm/systemSmartContracts/esdt.go

```diff
@@ -795,6 +805,206 @@ func (e *esdt) controlChanges(args *vmcommon.ContractCallInput) vmcommon.ReturnC
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+	token, returnCode := e.basicOwnershipChecks(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	address := args.Arguments[1]
+	esdtRole, isNew := getRolesForAddress(token, address)
+	if isNew {
+		e.eei.AddReturnMessage("address does not have special role")
+		return vmcommon.UserError
+	}
+
+	for _, arg := range args.Arguments[2:] {
+		index, exist := doesRoleExist(esdtRole, arg)
+		if !exist {
+			e.eei.AddReturnMessage("role does not exist for address")
+			return vmcommon.UserError
+		}
+
+		copy(esdtRole.Roles[index:], esdtRole.Roles[index+1:])
+		esdtRole.Roles[len(esdtRole.Roles)-1] = nil
+		esdtRole.Roles = esdtRole.Roles[:len(esdtRole.Roles)-1]
+	}
+
+	err := e.sendRoleChangeData(args, core.BuiltInFunctionUnSetESDTRole)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	err = e.saveToken(args.Arguments[0], token)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (e *esdt) sendRoleChangeData(args *vmcommon.ContractCallInput, builtInFunc string) error {
+	esdtSetRoleData := builtInFunc + "@" + hex.EncodeToString(args.Arguments[0])
+	for _, arg := range args.Arguments[2:] {
+		esdtSetRoleData += "@" + hex.EncodeToString(arg)
+	}
+
+	err := e.eei.Transfer(args.Arguments[1], e.eSDTSCAddress, big.NewInt(0), []byte(esdtSetRoleData), 0)
+	return err
+}
+
+func (e *esdt) getAllAddressesAndRoles(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.Arguments) != 1 {
+		e.eei.AddReturnMessage("needs 1 argument")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		e.eei.AddReturnMessage("callValue must be 0")
+		return vmcommon.UserError
+	}
+
+	token, err := e.getExistingToken(args.Arguments[0])
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	for _, esdtRole := range token.SpecialRoles {
+		e.eei.Finish(esdtRole.Address)
+		for _, role := range esdtRole.Roles {
+			e.eei.Finish(role)
+		}
+	}
+
+	return vmcommon.Ok
+}
+
+func checkDuplicates(arguments [][]byte) error {
+	mapArgs := make(map[string]struct{})
+	for _, arg := range arguments {
+		_, found := mapArgs[string(arg)]
+		if !found {
+			mapArgs[string(arg)] = struct{}{}
+		}
+		return vm.ErrDuplicatesFoundInArguments
+	}
+
+	return nil
+}
+
+func validateRoles(arguments [][]byte) error {
+	err := checkDuplicates(arguments)
+	if err != nil {
+		return err
+	}
+
+	for _, arg := range arguments {
+		err = isSpecialRoleValid(string(arg))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func isSpecialRoleValid(argument string) error {
+	switch argument {
+	case core.ESDTRoleLocalMint:
+		return nil
+	case core.ESDTRoleLocalBurn:
+		return nil
+	default:
+		return vm.ErrInvalidArgument
+	}
+}
+
+func doesRoleExist(esdtRoles *ESDTRoles, role []byte) (int, bool) {
```

Comment:
We can rename the function to "getRoleIndex" and remove the bool output. There is a common knowledge thing that a negative index means "not found".

Reply:
done